Marlin/planner.cpp

```diff
@@ -601,26 +591,47 @@ void Planner::check_axes_activity() {
 #endif // PLANNER_LEVELING
 
 /**
- * Planner::buffer_line
+ * Add a new linear movement to the buffer.
+ * The target is cartesian.
+ *
+ *  target   - x,y,z,e CARTESIAN target in mm
+ *  fr_mm_s  - (target) speed of the move (mm/s)
+ *  extruder - target extruder
+ */
+void Planner::buffer_line_kinematic(const float target[NUM_AXIS], float fr_mm_s, const uint8_t extruder) {
+  const float* t=target;
+  #if PLANNER_LEVELING
+    float leveled[NUM_AXIS];
+    memcpy(leveled, target, sizeof(leveled));
+    apply_leveling(leveled);
+    t=leveled;
+  #endif
```

Comment:
That said, let's see if it's possible to avoid this `memcpy` (and whether it makes sense to do so). First, consider the XYZE arguments to `Planner::buffer_line`. It seems that these could be reduced to two arguments: A pointer to an XYZ array, plus one E argument. However, to avoid altering the original variables, a copy has to be made anyway. The stack copy is only supplanted by the `memcpy`. The position pointer argument used by `buffer_line_kinematic` probably adds one pointer-push worth of overhead. Or, none if it's passed in a register.

And, "why?" –you may ask– do we bother with `ARG_X`, `ARG_Y`, `ARG_Z`? For one thing, each `float` is 4 bytes, while references are only 2. So references are favored. But unfortunately, the compiler doesn't treat `const &` as immutable. If a function alters the value, the original variable will be changed. So, when the planner isn't doing leveling, some cycles can be saved using the refs. The `buffer_line` and `set_position_mm` functions get copies of the original variables.

In my cursory analysis, I don't see that there have to be any added cycles.